### PR TITLE
Add basic hostname validation on Windows

### PR DESCRIPTION
**Pull request by tas50**
alphanumeric, -, . or _
Starts with an alphanumeric

We could get a lot more complex here and that might not be a bad idea, but this is probably a step in the right direction.

Signed-off-by: Tim Smith <tsmith@chef.io>

### CODE REVIEW

**Comment by lamont-granquist** [comment]
hrm, isn't this one of those things that we should typically let the underlying operating system just fail on and surface the error?  similar to the username and group regexps?

**Comment by tas50** [comment]
It might fail on Linux, but it's pretty clear you can pass bogus data to windows the way we're doing it.

**Comment by tas50** [comment]
At the very least we're going to have to have some Windows specific regex because we'll just reboot forever trying to set the hostname

**Comment by lamont-granquist** [comment]
Yeah that might be better to clean up windows since it explodes the server if you don't.

**Comment by tas50** [comment]
Round 2:

- Moves ohai into non-windows only since we don't use it on windows
- Raises if you put a period in the hostname on windows

File: lib/chef/resource/hostname.rb [hunk 1]
```diff
@@ -203,8 +203,9 @@ def updated_ec2_config_xml
           end
 
         else # windows
+          raise "Windows hostnames cannot contain a period." if new_resource.hostname.match?(/./)
 
-        # suppress EC2 config service from setting our hostname
```
Comment (tas50) [inline lib/chef/resource/hostname.rb]: rubocop totally fails to align comments btw. it's quite odd

**Comment by stuartpreston** [comment]
Here’s the reference I use for naming on Windows: https://support.microsoft.com/en-gb/help/909264/naming-conventions-in-active-directory-for-computers-domains-sites-and

This PR looks like a good gate stop though to catch the most likely cause of failure.

**Comment by tas50** [comment]
@stuartpreston I started with that doc and I still think we could benefit from a nice validation here, but it would potentially get a bit tricky since it's different on windows vs *nix

**Pull request merged**